api/bucketsd/service.go

```diff
@@ -970,16 +973,175 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (res *pb.
 	}
 	dbToken, _ := thread.TokenFromContext(ctx)
 
-	destPath := cleanPath(req.Path)
-	bootCid, err := cid.Decode(req.Cid)
+	if isEqualPath(req.Path, req.Destination) {
+		return &pb.MovePathResponse{}, nil
+	}
+
+	cleanPth, err := parsePath(req.Path)
 	if err != nil {
-		return nil, fmt.Errorf("invalid remote cid: %v", err)
+		return nil, err
+	}
+	cleanDst, err := parsePath(req.Destination)
+	if err != nil {
+		return nil, err
 	}
 
 	lck := s.Semaphores.Get(buckLock(req.Key))
 	lck.Acquire()
 	defer lck.Release()
 
+	buck, pth, err := s.getBucketPath(ctx, dbID, req.Key, cleanPth, dbToken)
+	if err != nil {
+		return nil, err
+	}
+	n, err := s.getNodeAtPath(ctx, pth, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.setPath(ctx, &pb.SetPathRequest{
+		Key:  req.Key,
+		Path: cleanDst,
+		Cid:  n.Cid().String(),
+	}, dbID, dbToken, cleanDst)
+	if err != nil {
+		return nil, fmt.Errorf("error setting path: %v", err)
+	}
+	if buck.IsPrivate() {
+		meta, _, ok := buck.GetMetadataForPath(cleanPth, false)
+		if !ok {
+			return nil, fmt.Errorf("error getting metadata")
+		}
+		buck.SetMetadataAtPath(cleanDst, meta)
+		if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+			return nil, fmt.Errorf("verifying bucket update: %v", err)
+		}
+	}
+
+	// if moving root into subdir, rm .textileseed from subdir
+	if cleanPth == "" {
+		filepath := gopath.Join(cleanDst, buckets.SeedName)
+		_, err = s.removePath(ctx, dbID, dbToken, &pb.RemovePathRequest{
+			Key:  req.Key,
+			Path: filepath,
+		}, filepath)
+
+		if err != nil {
+			return nil, fmt.Errorf("error cleaning path: %v", err)
+		}
+	}
+
+	// a nice absolute path to skip Rel errors
+	ipfsPath := gopath.Join("/ipfs", n.Cid().String(), cleanPth)
+	ipfsDest := gopath.Join("/ipfs", n.Cid().String(), cleanDst)
+
+	// If "a/b" => "a/" no cleanup is needed
+	parent, err := isSubPath(ipfsDest, ipfsPath)
+	if err != nil {
+		return nil, fmt.Errorf("path cleanup failed: %v", err)
+	}
+	if parent == true {
+		return &pb.MovePathResponse{}, nil
+	}
+
+	// If "a/" => "a/b" cleanup each leaf in "a" that isn't "b" (skipping .textileseed)
+	subdir, err := isSubPath(ipfsPath, ipfsDest)
+	if err != nil {
+		return nil, fmt.Errorf("path cleanup failed: %v", err)
+	}
+	if subdir == true {
+		item, err := s.ListPath(ctx, &pb.ListPathRequest{
```

Comment:
Using this here is nice since it encapsulates the encrypted vs. non-encrypted handling, but will make the method slower by doing more roundtrips to threads. We could leave it for this first pass though.